File: ui.c
#include "ui.h"

#include "card.h"
#include "range.h"

#include <stdio.h>

static Position read_position(void);
static double read_non_negative_double(const char *prompt);
static double read_probability(const char *prompt);
static int read_int_in_range(const char *prompt, int min, int max);
static HoleCards read_hole_cards(void);
static void clear_input_buffer(void);

static Position read_position(void) {
    int choice = 0;

    while (1) {
        printf("Hero position:\n");
        printf("1. UTG\n");
        printf("2. HJ\n");
        printf("3. CO\n");
        printf("4. BTN\n");
        printf("5. SB\n");
        printf("6. BB\n");
        printf("Choose position: ");

        if (scanf("%d", &choice) != 1) {
            clear_input_buffer();
            printf("Invalid input. Please enter a number.\n\n");
            continue;
        }

        if (choice >= 1 && choice <= 6) {
            return (Position)(choice - 1);
        }

        printf("Invalid position. Please choose 1 to 6.\n\n");
    }
}

static double read_non_negative_double(const char *prompt) {
    double value = 0.0;

    while (1) {
        printf("%s", prompt);

        if (scanf("%lf", &value) != 1) {
            clear_input_buffer();
            printf("Invalid input. Please enter a number.\n\n");
            continue;
        }

        if (value >= 0.0) {
            return value;
        }

        printf("Value cannot be negative.\n\n");
    }
}

static double read_probability(const char *prompt) {
    double value = 0.0;

    while (1) {
        printf("%s", prompt);

        if (scanf("%lf", &value) != 1) {
            clear_input_buffer();
            printf("Invalid input. Please enter a number.\n\n");
            continue;
        }

        if (value >= 0.0 && value <= 1.0) {
            return value;
        }

        printf("Probability must be between 0 and 1.\n\n");
    }
}

static int read_int_in_range(const char *prompt, int min, int max) {
    int value = 0;

    while (1) {
        printf("%s", prompt);

        if (scanf("%d", &value) != 1) {
            clear_input_buffer();
            printf("Invalid input. Please enter a number.\n\n");
            continue;
        }

        if (value >= min && value <= max) {
            return value;
        }

        printf("Value must be between %d and %d.\n\n", min, max);
    }
}

static HoleCards read_hole_cards(void) {
    char first_text[8];
    char second_text[8];
    HoleCards hand;

    while (1) {
        printf("Enter hero cards, e.g. Ah Ks: ");

        if (scanf("%7s %7s", first_text, second_text) != 2) {
            clear_input_buffer();
            printf("Invalid input. Please enter two cards.\n\n");
            continue;
        }

        if (!parse_hole_cards(first_text, second_text, &hand)) {
            printf("Invalid cards. Use rank 2-9,T,J,Q,K,A and suit c,d,h,s.\n\n");
            continue;
        }

        return hand;
    }
}

static void clear_input_buffer(void) {
    int ch = 0;

    while ((ch = getchar()) != '\n' && ch != EOF) {
    }
}

void run_main_menu(void) {
    int choice = 0;

    while (1) {
        printf("\nSix-max Preflop Decision System\n");
        printf("1. Solve preflop decision\n");
        printf("2. Input guide\n");
        printf("0. Exit\n");
        printf("Choose an option: ");

        if (scanf("%d", &choice) != 1) {
            clear_input_buffer();
            printf("Invalid input. Please enter a number.\n");
            continue;
        }

        if (choice == 1) {
            SolverInput input = read_solver_input();
            SolverResult result = solve_preflop_decision(input);
            print_solver_result(result);
        } else if (choice == 2) {
            print_input_guide();
        } else if (choice == 0) {
            printf("Goodbye.\n");
            return;
        } else {
            printf("Invalid option.\n");
        }
    }
}

SolverInput read_solver_input(void) {
    SolverInput input;

    printf("\nPreflop input\n");
    input.hero_position = read_position();
    input.hero_hand = read_hole_cards();
    input.pot_size = read_non_negative_double("Current pot size: ");
    input.call_amount = read_non_negative_double("Call amount: ");
    input.raise_amount = read_non_negative_double("Raise amount: ");
    input.fold_probability = read_probability("Estimated fold probability after raise (0-1): ");
    input.opponent_count = read_int_in_range("Opponent count (1-5): ", 1, 5);
    input.simulations = read_int_in_range("Monte Carlo simulations (100-1000000): ", 100, 1000000);

    return input;
}

void print_solver_result(SolverResult result) {
    printf("\nResult\n");
    printf("Equity: %.2f%%\n", result.equity * 100.0);
    printf("EV fold: %.3f\n", result.ev_fold);
    printf("EV call: %.3f\n", result.ev_call);
    printf("EV raise: %.3f\n", result.ev_raise);
    printf("Recommendation: %s\n", action_to_string(result.recommendation));

    if (result.explanation[0] != '\0') {
        printf("Explanation: %s\n", result.explanation);
    }
}

void print_input_guide(void) {
    printf("\nInput guide\n");
    printf("Cards use two characters: rank then suit.\n");
    printf("Ranks: 2 3 4 5 6 7 8 9 T J Q K A\n");
    printf("Suits: c clubs, d diamonds, h hearts, s spades\n");
    printf("Examples: Ah Ks, Td Tc, 7c 2d, As Kd\n");
    printf("Ten must be T, so use Th instead of 10h.\n");
    printf("Positions: UTG, HJ, CO, BTN, SB, BB.\n");
}

File: ui.py
from __future__ import annotations

from card import parse_hole_cards
from common import TABLE_POSITIONS, Action, HoleCards, PlayerAction, Position
from range_model import opening_range_summary, position_to_string
from solver import SolverInput, SolverResult, action_to_string, solve_preflop_decision


POSITIONS = TABLE_POSITIONS


def _read_int_range(prompt: str, minimum: int, maximum: int) -> int:
    while True:
        text = input(prompt).strip()
        try:
            value = int(text)
        except ValueError:
            value = minimum - 1

        if minimum <= value <= maximum:
            return value

        print(f"Invalid input. Please enter an integer from {minimum} to {maximum}.")


def _read_float_min(prompt: str, minimum: float) -> float:
    while True:
        text = input(prompt).strip()
        try:
            value = float(text)
        except ValueError:
            value = minimum - 1.0

        if value >= minimum:
            return value

        print(f"Invalid input. Please enter a number greater than or equal to {minimum:.2f}.")


def _read_hole_cards() -> HoleCards:
    while True:
        first = input("Hero first card  (example Ah): ").strip()
        second = input("Hero second card (example Ks): ").strip()
        hand = parse_hole_cards(first, second)

        if hand is not None:
            return hand

        print("Invalid cards. Use rank-suit format like Ah, Ks, Tc. Duplicates and 10h are invalid.")


def _read_position() -> Position:
    print("\nHero position")
    print("1. UTG")
    print("2. HJ")
    print("3. CO")
    print("4. BTN")
    print("5. SB")
    print("6. BB")

    choice = _read_int_range("Choose hero position: ", 1, 6)
    return Position(choice - 1)


def _read_any_position(prompt: str) -> Position:
    print("1. UTG  2. HJ  3. CO  4. BTN  5. SB  6. BB")
    choice = _read_int_range(prompt, 1, 6)
    return Position(choice - 1)


def _read_action_record(position: Position) -> PlayerAction:
    print(f"\nAction for {position_to_string(position)}")
    print("1. Fold")
    print("2. Call")
    print("3. Raise")
    print("4. Check")
    choice = _read_int_range("Choose action: ", 1, 4)
    action = {
        1: Action.FOLD,
        2: Action.CALL,
        3: Action.RAISE,
        4: Action.CHECK,
    }[choice]
    amount = 0.0
    if action == Action.CALL:
        amount = _read_float_min("Call amount added in BB (0 to derive automatically): ", 0.0)
    elif action == Action.RAISE:
        amount = _read_float_min("Raise-to total in BB: ", 0.01)
    return PlayerAction(position=position, action=action, amount=amount)


def _read_action_history() -> tuple[PlayerAction, ...]:
    count = _read_int_range("Number of prior action records (0-30): ", 0, 30)
    records: list[PlayerAction] = []
    for index in range(count):
        print(f"\nPrior action {index + 1}")
        position = _read_any_position("Choose acting position: ")
        records.append(_read_action_record(position))
    return tuple(records)


def _read_candidate_raise_amounts() -> tuple[float, ...]:
    print("\nCandidate raise sizes")
    print("Enter each raise-to total, not the additional amount invested.")
    print("Use 0 to skip raise analysis for this spot.")
    count = _read_int_range("Number of raise sizes to compare (0-5): ", 0, 5)
    values: list[float] = []
    for index in range(count):
        amount = _read_float_min(f"Raise option {index + 1} amount in BB: ", 0.01)
        values.append(amount)
    return tuple(values)


def run_main_menu() -> None:
    while True:
        print()
        print("============================================")
        print("Texas Hold'em Preflop Decision System")
        print("============================================")
        print("1. New analysis")
        print("2. Input guide")
        print("3. Exit")

        choice = _read_int_range("Choose an option: ", 1, 3)

        if choice == 1:
            solver_input = read_solver_input()
            result = solve_preflop_decision(solver_input)
            print_solver_result(result)
        elif choice == 2:
            print_input_guide()
        else:
            return


def read_solver_input() -> SolverInput:
    print("\nNew decision analysis")
    print("---------------------")

    hero_position = _read_position()
    hero_hand = _read_hole_cards()
    print("\nCurrent decision point")
    pot_size = _read_float_min("Current pot size in BB: ", 0.0)
    current_bet = _read_float_min("Current highest total bet in BB: ", 0.0)
    hero_contribution = _read_float_min("Hero amount already invested in BB: ", 0.0)
    call_amount = max(0.0, current_bet - hero_contribution)
    print(f"Amount Hero must call: {call_amount:.2f} BB")
    candidate_raise_amounts = _read_candidate_raise_amounts()
    active_opponent_count = _read_int_range("Active opponent count (0-5): ", 0, 5)
    table_actions = _read_action_history()
    simulations = _read_int_range("Simulation count (1-1000000): ", 1, 1_000_000)
    primary_raise_amount = candidate_raise_amounts[0] if candidate_raise_amounts else 0.0

    return SolverInput(
        hero_position=hero_position,
        hero_hand=hero_hand,
        pot_size=pot_size,
        call_amount=call_amount,
        raise_amount=primary_raise_amount,
        simulations=simulations,
        table_actions=table_actions,
        active_opponent_count=active_opponent_count,
        candidate_raise_amounts=candidate_raise_amounts,
        hero_contribution=hero_contribution,
        current_bet=current_bet,
    )


def print_solver_result(result: SolverResult) -> None:
    print("\nResult")
    print("------")
    print(f"Hero position: {position_to_string(result.hero_position)}")
    if result.table_actions:
        print("Entered actions:")
        for record in result.table_actions:
            print(
                f"  {position_to_string(record.position)} "
                f"{action_to_string(record.action)}"
                f"{f' +{record.amount:.2f}' if record.amount > 0.0 and record.action == Action.RAISE else ''}"
                f"{f' {record.amount:.2f}' if record.amount > 0.0 and record.action != Action.RAISE else ''}"
            )
    else:
        print("Entered actions: none")
    print(f"Estimated opponents: {result.opponent_count}")
    print(f"Hand class:    {result.hand_class.name}")
    print(f"In open range: {'yes' if result.range_frequency.open_frequency > 0.0 else 'no'}")
    print(f"Open freq:     {result.range_frequency.open_frequency * 100:.0f}%")
    print(f"Call freq:     {result.range_frequency.call_frequency * 100:.0f}%")
    print(f"Raise freq:    {result.range_frequency.raise_frequency * 100:.0f}%")
    print(f"Auto fold prob: {result.fold_probability * 100:.2f}%")
    print()
    print(f"Simulations: {result.equity_result.simulations}")
    print(f"Wins:        {result.equity_result.wins} ({result.equity_result.win_rate * 100:.2f}%)")
    print(f"Ties:        {result.equity_result.ties} ({result.equity_result.tie_rate * 100:.2f}%)")
    print(f"Losses:      {result.equity_result.losses} ({result.equity_result.loss_rate * 100:.2f}%)")
    print(f"Equity:      {result.equity:.4f}")
    print()
    print(f"EV fold:     {result.ev_fold:.4f}")
    print(f"EV check:    {result.ev_check:.4f}")
    print(f"EV call:     {result.ev_call:.4f}")
    print(f"EV raise:    {result.ev_raise:.4f}"
          f"{f' at {result.best_raise_amount:.2f} BB' if result.best_raise_amount > 0.0 else ''}")
    print("Action EVs:")
    for action_ev in result.action_evs:
        amount = f" {action_ev.amount:.2f} BB" if action_ev.amount > 0.0 else ""
        print(f"  {action_to_string(action_ev.action)}{amount}: {action_ev.ev:+.4f}")
    print()
    print(f"Recommendation: {action_to_string(result.recommendation)}")
    print(f"Explanation:    {result.explanation}")


def print_input_guide() -> None:
    print("\nInput guide")
    print("-----------")
    print("Card format: rank followed by suit.")
    print("Ranks: 2 3 4 5 6 7 8 9 T J Q K A")
    print("Suits: c=clubs, d=diamonds, h=hearts, s=spades")
    print("Valid examples: Ah, Ks, Qd, Tc, 7c")
    print("Invalid examples: 10h, Kx, ZZ, duplicated cards like Ah Ah")
    print()
    print("In a new analysis, enter Hero position and cards first.")
    print("Then enter the current pot, current highest bet, Hero contribution, raise-to totals, active opponents, and prior actions.")
    print_opening_ranges()


def print_opening_ranges() -> None:
    print("\nPosition opening range model")
    print("----------------------------")
    for pos in (Position.UTG, Position.HJ, Position.CO, Position.BTN, Position.SB, Position.BB):
        print(f"{position_to_string(pos)}: {opening_range_summary(pos)}")
